pallets/pallet-bonded-coins/src/traits.rs

```diff
@@ -0,0 +1,9 @@
+use sp_runtime::DispatchError;
+
+pub trait FreezeAccounts<AccountId, AssetId> {
+	type Error: Into<DispatchError>;
+
+	fn freeze(caller: &AccountId, who: &AccountId, asset_id: &AssetId) -> Result<(), Self::Error>;
```

Comment:
The fungibles traits typically don't ask who's trying to do the thing, they just get the thing done (e.g., a transfer from a to b). I think this trait should be designed the same way.

Reply:
Yes, in general I totally agree with you, but the `pallet_assets` does not provide any interface to change the freeze status of any account or asset. What I need to do is basically dispatch the corresponding tx on runtime level to change that. Therefore, I need an origin to dispatch the call. The tx is also permissioned. Only the freezer or the admin can change the state.

Reply:
Getting rid of the caller is not possible until there is a proper interface

Reply:
It should be possible by moving this bit out of the pallet implementation, and into the trait implementation:
```rust
				let freezer = T::Fungibles::freezer(target_currency_id.clone())
					// Should never fail. Either the freezer has been updated or it is the pool id.
					.ok_or_else(|| {
						// Todo: change log target
						log::error!("Freezer not found for currency id: {:?}", target_currency_id);
						Error::<T>::Internal
					})?;
```
This way the trait gets what feels to me like the proper level of abstraction, and shortcomings / implementation details of the assets pallet are dealt with in the trait implementation.

Reply:
Fair point: [2bb01ba](https://github.com/KILTprotocol/kilt-node/pull/770/commits/2bb01baea1178fb023c214a71e18473fa9eff2c3)